### PR TITLE
add oss-fuzz

### PR DESCRIPTION
**Pull request by cujomalainey**
See https://google.github.io/oss-fuzz/getting-started/new-project-guide/ for instruction on how to run code

Note this diff is functional but these commits need some reworking, will publish once commits are cleaned.

I am able now to produce a crash that is caused by bug in the firmware (rather than a bug in the harness).

TODO: produce seed corpus

Once related changes land in oss fuzz our project will be fuzzed on a daily basis automatically and have bugs filed against it. We should work to enable as much of the system as possible in the host build so the fuzzer can access it and fuzz it.

### CODE REVIEW

File: tools/oss-fuzz/fuzz_ipc.c [hunk 1]
```diff
@@ -22,7 +23,7 @@ int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size)
 	// copy the buffer to pre allocated buffer
 	struct sof_ipc_cmd_hdr *hdr = malloc(SOF_IPC_MSG_MAX_SIZE);
 
-	memcpy_s(hdr, Data, min(Size, SOF_IPC_MSG_MAX_SIZE));
+	memcpy_s(hdr, SOF_IPC_MSG_MAX_SIZE, Data, MIN(Size, SOF_IPC_MSG_MAX_SIZE));
```
Comment (lgirdwood) [inline tools/oss-fuzz/fuzz_ipc.c:26]: Should this be a PR for v1.6 fix ?

**Review by lgirdwood — COMMENTED** [review]
Good that we are picking up issues with the development fuzzer.

File: tools/oss-fuzz/fuzz_ipc.c (hunk 1 above)
Comment (cujomalainey) [reply tools/oss-fuzz/fuzz_ipc.c:26]: This is in the fuzzers stub itself, I will squash this fix into the original commit to add the fuzzers. It has no effect on the firmware.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by cujomalainey** [comment]
@singalsu i only built test the testbench, can you test this PR as well? I enable quite a few more modules here.

**Review by singalsu — APPROVED** [review]
All the tests run OK with this PR included. The code looks good to me as far as I can understand it. Also also thanks for static build! It was in my list of things needed.

**Comment by cujomalainey** [comment]
rebased and fixed formatting. the gdb init header is copied from xtensa arch and is extern there as well so im ignoring that error as well.

**Comment by cujomalainey** [comment]
fixed a bug that resulted in 2 different SOF contexts causing bad fuzzer bugs

**Comment by cujomalainey** [comment]
Looks like we have a decent amount of code that just assumes hw configs exist for things like DAIs, timers etc, im going to add some commits to stub out those configs for host platforms and add them to this PR so the fuzzer can stop segfaulting for silly reasons

**Comment by cujomalainey** [comment]
It looks like it might be worth splitting this up. The amount of sof struct properties that are not initialized that the fuzzer is finding is breaking a lot of the testing so I am having to stub out a lot more as well as the fact that the trace module is no compiled in (but is enabled) is causing issues as well. Might need to refactor some of the testbench tracing into that module

**Review by lgirdwood — APPROVED** [review]
@cujomalainey are we good to merge and fix up incrementally or do you still have more work to do prior to merge ?

**Comment by cujomalainey** [comment]
The fuzzer runs at this point, so yes i think we can work incrementally form here.

**Comment by marc-hb** [comment]
Note this PR added a spurious `.checkpatch-camelcase.git.` file, fixed in PR #3540 .

This was likely added by `git add -a`. Never `git add -a`, it's evil :-)

**Comment by cujomalainey** [comment]
@marc-hb no i do the much worse `git add .` thanks for catching that

**Pull request merged**